tests/fixtures/parsers/enhanced-parser3.js

```diff
@@ -0,0 +1,56 @@
+"use strict";
+
+const assert = require("assert");
+const ScopeManager = require("eslint-scope/lib/scope-manager");
+const Referencer = require("eslint-scope/lib/referencer");
+const Traverser = require("../../../lib/util/traverser");
+
+class EnhancedReferencer extends Referencer {
+    visitClass(node) {
+
+        // Visit decorators.
+        if (node.experimentalDecorators) {
+            for (const decorator of node.experimentalDecorators) {
+                this.visit(decorator);
+            }
+        }
+
+        // Do default.
+        super.visitClass(node);
+    }
+
+    Decorator(node) {
+        if (node.expression) {
+            this.visit(node.expression);
+        }
+    }
+}
+
+function analyzeScope(ast) {
+    const options = {
+        optimistic: false,
+        directive: false,
+        nodejsScope: false,
+        impliedStrict: false,
+        sourceType: "script",
+        ecmaVersion: 6,
+        childVisitorKeys: null,
+        fallback: Traverser.getKeys
+    };
+    const scopeManager = new ScopeManager(options);
+    const referencer = new EnhancedReferencer(options, scopeManager);
+
+    referencer.visit(ast);
+
+    return scopeManager;
+}
+
+exports.parse = code => {
```

Comment:
I definitely agree with @not-an-aardvark here, and I have started working based on this assumption in my babel-eslint WIP